### DISCUSSION TITLE
Stateful parser

### DISCUSSION THREAD

**Discussion by Truedrog** [j-mie6/Parsley#178] [Q&A] [answered]
So here I am again with the question about statefull parser :)
Suppose we have some datatype
```

 case class Root(
                      items: List[Item],
                      comments: String)
```

suppose I have a parser which parse items for my root, and I want to update my Root with it after each parse
something likes this
`rootP: Parsley[Root] = many(addRootItemP) >> eof  >> get`

```
addRootItemP: Parsley[Void] = 
  choice(
    itemp  >>= modify(addItem)
  )
```

In haskell using megaparsec this was easy, I had to pass in my user state and then run it with my parser.
Like this 

```
type Parser = StateT String (ParsecT Void Text Identity)

parser :: Parser String
parser = a <|> b
  where
    a = "foo" <$ put "branch A" <* empty
    b = get   <* put "branch B"

main :: IO ()
main = do
  let p            = runStateT parser "initial"
      Right (a, s) = runParser p "" ""
  putStrLn ("Result:      " ++ show a)
  putStrLn ("Final state: " ++ show s)
  ```
  
So my question is this.  Looking at the documentation, I need to use Reg interface with it's methods(put, modify and get)
ok I can achive this if all my parsers in one file.
Like this 
```
case class Item(date: LocalDate, description: String)

case class Root(items: List[Item])

val r = Reg.make[Root]

val p =
  (datep, lookAhead(spacenonewline <|> newline).void, descriptionp).zipped[Item] { case (date, _, description) => Item(date, description) }
val addItemP = choice(
  p.flatMap(item => r.modify(root => root.copy(item +: root.items))),
  emptyorcommentlinep.void,
  multilinecommentp.void)
val parser = r.put(Root(List.empty)) *> many(addItemP) *> r.get
println(parser.parse("2022-03-27 change22\n2022-03-27 change23"))
```
What if I have addItemP or any other parser that might need a register in another file?
Pass it around as argument? 
Please apologies for long post!

**Comment by j-mie6** [comment]
Just use it in the other file!

Registers are bound to one specific _complete_ parser, but not to a file. You can just use the register (or the parsers) you've defined wherever. As long as it's used solely within one _.parse and not across two different ones it's fine. You can, of course pass it as an argument if you wanted, but closing over it is the preferred way (since all recursion must be knot-tied anyway). Does that make sense?

**Reply by j-mie6** [reply]
And p.s. you shouldn't need flatMap for that, the register API deliberately exposes versions that take parsers as arguments: so-called applicative State Vs monadic state

**Reply by j-mie6** [reply]
Also, I'm actually not entirely sure what happens if you do use a register in two different complete parsers: might cause problems, might not. Ultimately the registers themselves don't hold state, they just refer to a place that does within the internal machinery of a parser.

**Reply by Truedrog** [reply]
Well I am not gonna use another parser, just this one. By parser I mean _.parse call. 
>Ultimately the registers themselves don't hold state

So there is no convenient way to statefully update my structure in the process of parsing?

**Reply by j-mie6** [reply]
Sorry, misunderstanding: registers do allow you to push state through a parser, which is exactly what you want (it's the one-to-one replacement for the user state in megaparsec, just more flexible). It's just that _techinically_ when considering registers interactions with other parsers, the registers themselves don't hold state: they are effectively just indices into the internal states of each individual parser -- they are pointing to the state, as opposed to holding it themselves, names for state if you will. When considering one parser in isolation, the distinction is meaningless and you can think of it being that the register itself has the state inside

**Reply by j-mie6** [reply, marked as answer]
The key is that if a register were used in two parsers (which is undefined behaviour), it's not like the two parsers can communicate with each other via this register: they both have independent state because they are independent `_.parse` calls. Like I said, a distinction that doesn't matter when you're only working with one `_.parse`
